plugins/global-search/src/utils/indexer/indexer.ts

```diff
@@ -0,0 +1,183 @@
+
+                const [text, name] = await Promise.all([this.getNodeText(node), getNodeName(node)])
+
+                batch.push({
+                    id: node.id,
+                    type: node.__class,
+                    name,
+                    text,
+                    node,
+                    rootNode,
+                    rootNodeName,
+                    rootNodeType: rootNode.__class,
+                })
+
+                if (batch.length === this.batchSize) {
+                    yield batch
+                    batch = []
+                }
+            }
+        }
+
+        if (batch.length > 0) {
+            yield batch
+        }
+    }
+
+    private async *crawlCollections(collections: readonly Collection[]): AsyncGenerator<IndexEntry[]> {
+        let batch: IndexEntry[] = []
+
+        for (const collection of collections) {
+            const [fieldNames, items] = await Promise.all([collection.getFields(), collection.getItems()])
+            const fieldNameMap = new Map(fieldNames.map(f => [f.id, f.name]))
+
+            for (const item of items) {
+                const fields = Object.fromEntries(
+                    Object.entries(item.fieldData).flatMap(([key, field]) => {
+                        const finalKey = fieldNameMap.get(key) ?? key
+                        switch (field.type) {
+                            case "string":
+                                return [[finalKey, field.value]]
+                            case "formattedText":
+                                return [[finalKey, stripMarkup(field.value)]]
+                        }
+                        return []
+                    })
+                )
+
+                batch.push({
+                    id: item.id,
+                    type: "CollectionItem",
+                    collectionItem: item,
+                    rootNode: collection,
+                    rootNodeName: collection.name,
+                    rootNodeType: "Collection",
+                    fields,
+                    slug: item.slug,
+                })
+            }
+
+            if (batch.length === this.batchSize) {
+                yield batch
+                batch = []
+            }
+        }
+
+        if (batch.length > 0) {
+            yield batch
+        }
+    }
+
+    async start() {
+        try {
+            const [pages, components] = await Promise.all([
+                framer.getNodesWithType("WebPageNode"),
+                framer.getNodesWithType("ComponentNode"),
+            ])
+
+            this.abortRequested = false
+            this.eventEmitter.emit("started", undefined)
+
+            for await (const batch of this.crawlNodes([...pages, ...components])) {
+                if (this.abortRequested) break
+                this.upsertEntries(batch)
+            }
+
+            const collections = await framer.getCollections()
+
+            for await (const batch of this.crawlCollections(collections)) {
```

Comment:
I think out of scope for this PR, but I wonder if you are considering something to make this not freeze the UI on huge projects? 

Idk maybe like something with `requestIdleCallback` or some kind of scheduling? A worker probs will be more overhead because everything has to go through `framer` anyway